shimmingtoolbox/optimizer/lsq_optimizer.py

```diff
@@ -109,12 +109,13 @@ def _residuals_mse(self, coef, unshimmed_vec, coil_mat, factor):
                             avoid positive directional linesearch
 
         Returns:
-            numpy.ndarray: Residuals for least squares optimization -- equivalent to flattened shimmed vector
+            float: Residuals for least squares optimization -- equivalent to flattened shimmed vector
         """
-
+        # Old one was : np.mean((unshimmed_vec + np.sum(coil_mat * coef, axis=1, keepdims=False))**2) / factor + \
+        #                (self.reg_factor * np.mean(np.abs(coef) / self.reg_factor_channel))
         # MSE regularized to minimize currents
-        return np.mean((unshimmed_vec + np.sum(coil_mat * coef, axis=1, keepdims=False)) ** 2) / factor + \
-               (self.reg_factor * np.mean(np.abs(coef) / self.reg_factor_channel))
+        inner = unshimmed_vec + np.matmul(coil_mat, coef)
+        return inner.dot(inner) / len(inner) / factor + np.abs(coef).dot(self.reg_vector)
```

Comment:
- Use `@` operator instead of `np.matmul` for consistency within the code.
- Rename inner to something more appropriate
- Does making it a one-liner improve the speed (`len(inner) == len(unshimmed_vec)`)
- Are you saying that` inner.dot(inner) / len(inner)` is equal and faster to `np.mean(inner ** 2)`?
- Can you show how the regularization penalty term that you now have is the same as what was there before

Reply:
For the three first points, I changed it. 
Then, for the second point : 
The demonstration that inner.dot(inner) / len(inner) is equal to np.mean(inner**2) can be done with the following mathematical law :  
The scalar product (dot product) of two vectors is equal to the sum of the product of their respectives elements.
Let inner be a vector of dimension n, we have :
inner.dot(inner) = inner[0]*inner[0] + inner[1]*inner[1] + ... + inner[n-1]*inner[n-1]
Then the mean of inner**2 is equal :
np.mean(inner**2) = (inner[0]**2 + inner[1]**2 + ... + inner[n-1]**2) / n
By dividing  inner.dot(inner) by n, we have :
inner.dot(inner) / n = (inner[0]**2 + inner[1]**2 + ... + inner[n-1]**2) / n
This is the proof that inner.dot(inner) / n is equal at np.mean(inner**2).
Then for your last point 
This can be proved by applying the definition of the multiplication of two vectors of size n $n$, and that  $\mathbf{a}$ is the vector $\left| coef \right|$ and $\mathbf{b}$ is the vector $\frac{reg_{factor}}{n reg_{factor_channel}}$ . 

Where $a_i$ and $b_i$ are the corresponding elements of each vector.

By applying the definition of the multiplication of vectors, we have:

$\mathbf{a}\cdot\mathbf{b}=a_{1}b_{1}+a_{2}b_{2}+\cdots+a_{n}b_{n}$

By using the notation of the summation, this can be written like:

$\sum_{i=1}^{n} a_{i}b_{i}$

By switching the expression of $\mathbf{a}$ and $\mathbf{b}$, we have :

$\sum_{i=1}^{n} \left| coef_i \right| \frac{reg_{factor}}{n reg_{factor_channel}}$

And by reorganising, we have :

$\frac{reg_{factor}}{n reg_{factor_channel}} \sum_{i=1}^{n} \left| coef_i \right|$

With the definition of the mean, we can written the sum as:

$n \frac{1}{n} \sum_{i=1}^{n} \left| coef_i \right|$

If we compare this expression with $reg_factor * np.mean(np.abs(coef) / reg_factor_channel)$, we can see that they are equivalent.

Then if you compare the time of the first version and the one that I suggest, you can see that the speed of the function is decreased by a lot.